core/src/jsMain/kotlin/dev/fritz2/remote/auth.kt

```diff
@@ -0,0 +1,104 @@
+package dev.fritz2.remote
+
+import dev.fritz2.binding.storeOf
+import kotlinx.coroutines.CompletableDeferred
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.map
+
+/**
+ * Special [Interceptor] to use in at [http] API to provide an authentication
+ * for every request. The type-parameter [P] represents the principal information,
+ * which contains all login-information needed to authenticate against the external HTTP API.
+ * The principal information is held inside and available as flow by calling [principal].
+ * To use this [Interceptor] you need to implement the [addAuthentication] method, where you
+ * get the principal information, when available, to append it to your requests.
+ * To start the authentication process you also need to implement the [authenticate] method in which
+ * you specify what is needed to authenticate the user (e.g. open up a login modal).
+ * When your authentication process is done you have to call the [complete] function and set your
+ * principal. Then all requests, that were made in the meantime, getting re-executed with the additional
+ * authentication information provided by the [addAuthentication] method.
+ * When the user logs out you have to call the [clear] function to clear all authentication information.
+ */
+abstract class Authentication<P> : Interceptor {
+
+    private val principalStore = storeOf<P?>(null)
+
+    private var state: CompletableDeferred<P>? = null
+
+    /**
+     * List of HTTP-Status-Codes forcing an authentication.
+     * Defaults are 401 (unauthorized) and 403 (forbidden).
+     */
+    open val statusCodesEnforcingAuthentication: List<Int> = listOf(401, 403)
```

Comment:
As the only operation applied onto this is looking up a specific code, this should be a `Set` and not a `List`. I also cannot imagine some use case the sequential order will ever be relevant.

Reply:
changed